### PR TITLE
Check CUDA runtime availability in MXNet when building Horovod

### PR DESCRIPTION
**Pull request by apeforest**
This PR performs two checks:
1) If HOROVOD_GPU_ALLREDUCE/BROADCAST/ALLGATHER is set but MXNet is not built with CUDA runtime, horovod installation should error out
2) If HOROVOD_GPU_ALLREDUCE/BROADCAST/ALLGATHER is not set but MXNet is built with CUDA runtime, horovod should be built with CUDA runtime.

### CODE REVIEW

**Comment by apeforest** [comment]
@alsrgv Added CUDA check per your suggestion. This PR should address your TODO 1) and 3) in https://github.com/uber/horovod/blob/buildkite/horovod/mxnet/mpi_ops.cc#L294. Please help to review.

**Comment by apeforest** [comment]
@yuxihu @ctcyang Please also help to review. Thanks

**Comment by yuxihu** [comment]
Have you run the unittests on GPU instance?

File: setup.py [hunk 1]
```diff
@@ -622,6 +634,11 @@ def build_mx_extension(build_ext, options):
     build_ext.build_extension(mxnet_mpi_lib)
 
 
+def is_mx_cuda():
+    from mxnet import context
+    return context.num_gpus() > 0
```
Comment (alsrgv) [inline setup.py]: That will not work if we're building Horovod in Docker on the machine that does not have GPUs.  For PyTorch we had to implement dummy compilation to test if CUDA is enabled in the binary.  If MXNet does not offer diagnostic flag to tell if it was built with CUDA support, maybe we can also do a test compilation?

Comment (apeforest) [reply setup.py]: @alsrgv Could you elaborate on how you are going to test on the machine without GPUs? Do you mean you will build a docker image with CUDA and run it on a machine without GPU?

Comment (alsrgv) [reply setup.py]: The use case is to build a docker image with CUDA on the machine without GPU, but run on the machine with GPU.

**Comment by alsrgv** [comment]
Bump. @apeforest, I'm hoping we can get it in before the release.

**Comment by apeforest** [comment]
@alsrgv Got too busy with other projects last week. I will work on this tomorrow.

**Comment by alsrgv** [comment]
@apeforest, we've moved Horovod to its own organization, which requires DCO (dev certificate of origin).  Could you sign commits as per https://github.com/horovod/horovod/pull/810/checks?check_run_id=62082468?

File: horovod/mxnet/mpi_ops.cc [hunk 2]
```diff
@@ -290,8 +290,7 @@ extern "C" int horovod_mxnet_broadcast_async(NDArray* input, NDArray* output,
     DoBroadcast(input, output, root_rank, op_name, on_complete);
   };
 
-#if HAVE_CUDA && HOROVOD_GPU_BROADCAST != 'M'
-  ThrowIfError(common::CheckInitialized());
```
Comment (alsrgv) [inline horovod/mxnet/mpi_ops.cc]: We should keep this check here since it's not done in `DoBroadcastCudaOnCPU`.

File: setup.py [hunk 3]
```diff
@@ -622,6 +633,16 @@ def build_mx_extension(build_ext, options):
     build_ext.build_extension(mxnet_mpi_lib)
 
 
+def is_mx_cuda():
+    from mxnet import runtime
+    features = runtime.libinfo_features()
```
Comment (yuxihu) [inline setup.py]: This API is nice. Can we have an alternative approach to detect the cuda if the API is not there? In this way we do not lose Horovod support in the upcoming MXNet 1.4.0 release. Have you tried to detect the feature from libmxnet.so dependencies?

File: horovod/mxnet/mpi_ops.cc (hunk 2 above)
Comment (apeforest) [reply horovod/mxnet/mpi_ops.cc]: I have moved this into DoBroadcastCudaOnCPU to be consistent with other APIs.

**Comment by apeforest** [comment]
> Could you run tests on GPU instance?

Currently tests fail on GPU instances when HOROVOD_GPU_(ALLREDUCE|BROADCAST|ALLGATHER) is not set to NCCL. The reason is that the branch of DoXXXCudaOnCPU was never covered by current tests before (it was either HAVE_CUDA=0 or HOROVOD_GPU_ALLREDUCE=NCCL). I am working to fix these errors now.

**Comment by apeforest** [comment]
Close this PR for now and will resend once all tests are fixed.

**Comment by alsrgv** [comment]
@apeforest, any luck with the PR?  Tomorrow is the release day, would be lovely to merge in before that.

**Comment by apeforest** [comment]
@alsrgv The change is non-trivial. Even if I could push it in today, I think it's a little bit risky for the release tomorrow. What's the importance of this feature for the release and if so would it be possible to postpone the release by a few days?

**Comment by alsrgv** [comment]
@apeforest, I think we can release without it then, and follow up with 0.16.1 in a week or so.

**Pull request closed without merging**